### PR TITLE
[POAE7-2516] Next-gen compiler framework API

### PR DESCRIPTION
**Pull request by xieqi**
### What changes were proposed in this pull request?

initial commit for next-gen compiler framework API

### Why are the changes needed?

API for next-gen compiler

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Add a UT

### Which label does this PR belong to?

cider

### CODE REVIEW

**Review by winningsix — COMMENTED** [review]
Please add some comments for each key classes' responsibilities (incl. ```OpNode```, ```OpPipeline```,  ```Parser```, ```Transfomer``` and ```Translator```).

File: cider/exec/nextgen/OpPipeline.h [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+#include <vector>
+
+#include "exec/nextgen/OpNode.h"
+
+namespace cider::exec::nextgen {
+
+class OpPipeline {
+ public:
+  OpPipeline(const OpNodeVector& nodes) : nodes_(nodes) {}
```
Comment (Ma-Jian1) [inline cider/exec/nextgen/OpPipeline.h:33]: according to the usage in Parser, seems we should provide a move constructor

File: cider/exec/nextgen/Translator.h [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+class Context;
+class JITTuple;
+
+class Translator {
+ public:
+  Translator() = default;
+  virtual ~Translator() = default;
+
+  virtual void consume(Context* context, JITTuple* input) {}
```
Comment (Ma-Jian1) [inline cider/exec/nextgen/Translator.h:37]: reference may be better than pointer

File: cider/exec/nextgen/OpNode.h [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  OpNode() = default;
+  virtual ~OpNode() = default;
+
+  /// The name of the operator node
+  virtual const char* name() const = 0;
+
+ protected:
+  // input_;
+  // output_;
+  // schema_;
```
Comment (Ma-Jian1) [inline cider/exec/nextgen/OpNode.h:41]: where are expressions?

Comment (chaojun-zhang) [reply cider/exec/nextgen/OpNode.h:41]: I think the expressions member should be move inside the concreate  subclass

File: cider/exec/nextgen/OpNode.h [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+namespace cider::exec::nextgen {
+
+class OpNode {
+ public:
+  OpNode() = default;
+  virtual ~OpNode() = default;
+
+  /// The name of the operator node
+  virtual const char* name() const = 0;
```
Comment (chaojun-zhang) [inline cider/exec/nextgen/OpNode.h:36]: no  outputType?

File: cider/exec/nextgen/OpNode.h [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+};
+
+using OpNodePtr = std::shared_ptr<OpNode>;
+using OpNodeVector = std::vector<OpNodePtr>;
+
+class FilterNode : public OpNode {
+ public:
+  FilterNode() : OpNode() {}
+
+  const char* name() const override { return "FilterNode"; }
```
Comment (chaojun-zhang) [inline cider/exec/nextgen/OpNode.h:51]: no need filter expression in constructor?

File: cider/exec/nextgen/OpNode.h [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+ public:
+  OpNode() = default;
+  virtual ~OpNode() = default;
+
+  /// The name of the operator node
+  virtual const char* name() const = 0;
+
+ protected:
+  // input_;
+  // output_;
```
Comment (yma11) [inline cider/exec/nextgen/OpNode.h:40]: Do we still need `output_` node if it already has `input_`?

File: cider/exec/nextgen/Parser.h [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+#include "exec/nextgen/OpPipeline.h"
+#include "exec/template/RelAlgExecutionUnit.h"
+
+namespace cider::exec::nextgen {
+
+class Parser {
+ public:
+  // source--> filter -->sink
+  static std::unique_ptr<OpPipeline> toOpPipeline(const RelAlgExecutionUnit& eu) {
+    OpNodeVector nodes;
```
Comment (yma11) [inline cider/exec/nextgen/Parser.h:36]: `nodes = ConstructNodes(eu)`? So actually, we add an internal layer of `PlanNode `representation inside Cider and in future these nodes should be directly mapped to `Substrait::Rel` with `RelAlgExecutionUnit`. So do we need to design these nodes based on contents of corresponding substrait representation?

File: cider/exec/nextgen/Translator.h [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+ protected:
+  std::shared_ptr<OpNode> opNode_;
+  std::shared_ptr<Translator> successor_;
+};
+
+class FilterTranslator : public Translator {
+ public:
```
Comment (harborn) [inline cider/exec/nextgen/Translator.h:45]: nothing added in `FilterTranslator` ?

File: cider/exec/nextgen/OpNode.h (hunk 4 above)
Comment (xieqi) [reply cider/exec/nextgen/OpNode.h:41]: schema is for outputType

File: cider/exec/nextgen/OpNode.h (hunk 6 above)
Comment (xieqi) [reply cider/exec/nextgen/OpNode.h]: yep, will delete

File: cider/exec/nextgen/Translator.h [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+class Translator {
+ public:
+  Translator() = default;
+  virtual ~Translator() = default;
+
+  virtual void consume(const JITTuple& input, Context* context) {}
+
+ protected:
+  std::shared_ptr<OpNode> opNode_;
+  std::shared_ptr<Translator> successor_;
```
Comment (jikunshang) [inline cider/exec/nextgen/Translator.h:45]: `toTranslator` returns `std::unique_ptr<Translator>`, should we align them? 
And need a `setSuccessor` method or constrcutor will construct `successor_`

File: cider/exec/nextgen/Transformer.h [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+#include "exec/nextgen/OpPipeline.h"
+#include "exec/nextgen/Translator.h"
+
+namespace cider::exec::nextgen {
+
+/// \brief A transformer convert an OpPipeline to a translator
+class Transformer {
+ public:
+  static std::unique_ptr<Translator> toTranslator(const OpPipeline& pipeline) {
+    return std::make_unique<Translator>();
```
Comment (jikunshang) [inline cider/exec/nextgen/Transformer.h:36]: Any other class/API needed here? form pipeline to Translator. I suggest to list some even it's internal method.

File: cider/exec/nextgen/OpNode.h [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+  // input_;
+  // schema_;
+};
+
+using OpNodePtr = std::shared_ptr<OpNode>;
+using OpNodeVector = std::vector<OpNodePtr>;
+
+class FilterNode : public OpNode {
+ public:
+  FilterNode() : OpNode() {}
```
Comment (Ma-Jian1) [inline cider/exec/nextgen/OpNode.h:53]: FilterNode must contain expressions, the default construtor should be deleted.

File: cider/exec/nextgen/OpNode.h [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+
+  FilterNode(std::vector<Analyzer::Expr*>& filter) : filter_(std::move(filter)) {}
+
+  const char* name() const override { return "FilterNode"; }
+
+ private:
+  std::vector<Analyzer::Expr*> filter_;
```
Comment (Ma-Jian1) [inline cider/exec/nextgen/OpNode.h:60]: `std::vector<std::shared_ptr<Analyzer::Expr>>` should be easier to manage the memory

File: cider/exec/nextgen/OpPipeline.h [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+class OpPipeline {
+ public:
+  OpPipeline() = default;
+  OpPipeline(const OpNodeVector& nodes) : nodes_(nodes) {}
+  virtual ~OpPipeline() = default;
+
+  /// \brief Register an operator node in this pipeline
+  ///
+  /// \input opNode The operator node to add to the pipeline
+  void registerOpNode(const OpNodePtr opNode);
```
[the remark this answers is not in the document]
Comment (winningsix) [reply cider/exec/nextgen/OpPipeline.h:42]: Same question here. IMO, ```register``` is more for defining a new external type. ```add``` or ```append``` is more straight forward.

[the remark this answers is not in the document]
Comment (xieqi) [reply cider/exec/nextgen/OpPipeline.h]: > When will this API be used? always add to the end of the pipeline?
yes always add to the end of the pipeline, you can build the pipeline with all the opnodes or build an empty pipeline then add opnode to the pipeline.

**Review by bigPYJ1151 — APPROVED** [review]

**Review by winningsix — APPROVED** [review]

**Pull request merged**